### DISCUSSION TITLE
Where does the information rendered by `bundle info` come from?

### DISCUSSION THREAD

**Discussion by caasssandre** [ruby/rubygems#4949] [Q&A] [answered]
Hi All,

I have recently come across some behaviour that I am struggling to explain, and I hope I might find some enlightenment here.

Steps to reproduce:

1. In a Rails app, find the directory where gems are being installed. In mine, it is there: `/bundle/ruby/2.7.0/gems`
2. List the gems and make note of what the first one is, mine is called `addressfinder`
3. Run `bundle info your_gem_name`, you should get something like that: <img width="537" alt="Screen Shot 2021-10-01 at 2 43 23 PM" src="https://user-images.githubusercontent.com/41363645/135554400-92103f34-b6d2-4fee-813a-d85b08772d9f.png">
4. Remove the gem from the directory (use mv and move it to a temporary location so you can put it back later)
5. Run `bundle info your_gem_name` again, you will see no difference to the first time
6. Now list all the gems again, you will see that the gem you moved is not there anymore but `bundle info` still tells you that the path to the gem is unchanged.
7. If you try to run your app (assuming it is using the gem) it will be broken
8. If you try to run `bundle install` nothing will happen, bundler seems to think that the gem is installed, even though it isn't there anymore.

The only way I found to get the app working, is to copy and paste the gem from another repo/branch and put it back in the directory where it belongs.

I am wondering how `bundle info` works, where the information rendered by the command comes from, and if there is a way to reinstall the gems from the `Gemfile.lock` ignoring the information from `bundle info`

Cheers 🙏

**Comment by deivid-rodriguez** [comment, marked as answer]
mmmm this seems mostly unexpected to me.

* I believe `bundle info` should be able to detect this a give a better message: https://github.com/rubygems/rubygems/blob/d01d6ca84011b05251b31d496f75f6aff8f677c4/bundler/lib/bundler/cli/info.rb#L46-L49.
* Also `bundle install` should verify whether (even if the lockfile is up to date and includes a correct resolution), you're missing any gems locally and trigger a reinstall in that case: https://github.com/rubygems/rubygems/blob/d01d6ca84011b05251b31d496f75f6aff8f677c4/bundler/lib/bundler/installer.rb#L275-L282 (the `@definition.missing_gems?` part should be verifying this).

Can you open a ticket with steps to repro? I'd like to know what we are missing.

Also, a workaround should be to pass `--force` to `bundle install`.

**Reply by caasssandre** [reply]
Thank you for your reply, I have created an issue [here](https://github.com/rubygems/rubygems/issues/4954)
